rustc_codegen_llvm: Adjust RISC-V inline assembly's clobber list

a4lg opened the pull request:
Despite that the `fflags` register (representing floating point exception flags) is stated as a flag register [in the reference](https://doc.rust-lang.org/reference/inline-assembly.html#r-asm.rules.preserved-registers), it's not
in the default clobber list of the RISC-V inline assembly and it would be better to fix it.

<!-- homu-ignore:start -->
# Related

*   rust-lang/rust#87581

@rustbot label +O-riscv
<!-- homu-ignore:end -->

jieyouxu commented:
r? compiler

nnethercote commented on compiler/rustc_codegen_llvm/src/asm.rs:
```diff
@@ -240,6 +240,8 @@ impl<'ll, 'tcx> AsmBuilderMethods<'tcx> for Builder<'_, 'll, 'tcx> {
                 InlineAsmArch::RiscV32 | InlineAsmArch::RiscV64 => {
                     constraints.extend_from_slice(&[
+                        "~{fflags}".to_string(),
+                        "~{frm}".to_string(),
```
The reference mentions `fflags` but does not mention `frm` or `vxrm`. Should it?

a4lg replied:
I'm not sure because `frm` and `vxrm` are rounding mode registers (the developer explicitly writes).
So, possible resolution might be opposite: removing them from the clobber list (the only reason I included `frm` in this list is because `vxrm` was already there).

taiki-e replied:
From the reference:

> - Floating-point exception flags in `fcsr` (`fflags`).
> - Vector extension state (`vtype`, `vl`, `vcsr`).

From RISC-V Unprivileged ISA V20250508 (Section "20.2. Floating-Point Control and Status Register" and "30.3.10. Vector Control and Status (vcsr) Register"):

- `fcsr` is `fflags` (bits 4:0) + `frm` (bits 7:5) + reserved bits.
- `vcsr` is `vxsat` (bit 0) + `vxrm` (bits 2:1) + reserved bits.

IIUC the presence of `vxrm` here matches the reference, but `frm` does not.

Also, changing the floating-point rounding mode in Rust usually causes immediate undefined behavior, so even if we marked `frm` as clobbered, it's unclear whether it can be used properly. (see also https://github.com/rust-lang/stdarch/pull/1478 and docs added in that.)

nnethercote replied:
Sounds like this PR shouldn't be merged, then?

a4lg replied:
@Amanieu 
I see, the ABI matter (I missed that).

So, the best solution here will be:
1.  Remove `frm` from the clobber register list (now)
2.  Add vector state sub-registers including `vxrm` (similar to "`fcsr` (`fflags`)") to the reference for clarification (not urgent)

I'll remove `frm` on the next version.

taiki-e commented:
cc @Amanieu

a4lg commented:
@nnethercote

I'll rewrite the commit message and I'll write a part of the draft v2 below (for clarity).

Despite that the `fflags` register is stated as a flag register in the reference, it's not in the clobber list of the RISC-V inline assembly and it would be better to fix it.  (cont.)

a4lg commented:
I'm investigating the LLVM source code for interactions between *subset registers* (e.g. `fflags` is a readable+writable mirror of certain bits of `fcsr`) and will adjust the clobber list based on the results.

### About adding `frm` to the list

If we don't have enough benefits having that, we have an option **NOT** to do it without causing any real problems (since this change will be backward-compatible and the only reason I initially added `frm` to the clobber list in this PR is for consistency).

### About removing `vxrm` from the list

If we do that, we technically make a backward-incompatible change.
Still, `vxrm` is only used by RISC-V vectors (RVV; based on dynamically-sized vectors like Arm SVE) and we are [not yet ready](https://rust-lang.github.io/rust-project-goals/2025h1/arm-sve-sme.html) for such vectors.  So, removing `vxrm` from the clobber list will be a safe option until we start actively supporting RVV through intrinsics and vector types.

a4lg committed:
rustc_codegen_llvm: Adjust RISC-V inline assembly's clobber list

Despite that the `fflags` register (representing floating point
exception flags) is stated as a flag register in the reference, it's not
in the default clobber list of the RISC-V inline assembly and it would
be better to fix it.

a4lg commented:
Removed `frm` from the clobber list and overhauled the commit message.

Amanieu commented:
We don't need to clobber vcsr since we already clobber vxrm and vxsat separately. There is nothing else in vcsr.

bors commented:
:pushpin: Commit 5ebdec5ac2908b0bae42adbe451beeadbe8fa5de has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 5ebdec5ac2908b0bae42adbe451beeadbe8fa5de -->
<!-- homu: {"type":"Approved","sha":"5ebdec5ac2908b0bae42adbe451beeadbe8fa5de","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
